File: ios/Sources/TaigiKeyboard/Overlays/SymbolData.swift
import Foundation

/// Symbol category for the symbol selection overlay
enum SymbolCategory: CaseIterable {
    case fullWidth
    case halfWidth
    case hiragana
    case katakana
    case kaomoji

    // i18n string key for the category tab label — resolved by DisplayLanguageStore under the
    // active display language (replaces the previously hardcoded Hanji label).
    var labelKey: StringKey {
        switch self {
        case .fullWidth: .symbolFullWidth
        case .halfWidth: .symbolHalfWidth
        case .hiragana: .symbolHiragana
        case .katakana: .symbolKatakana
        case .kaomoji: .symbolKaomoji
        }
    }

    var columnCount: Int {
        switch self {
        case .kaomoji: 2
        default: 6
        }
    }

    var fontSize: CGFloat {
        switch self {
        case .kaomoji: 14
        default: 20
        }
    }
}

/// Symbol data for the symbol selection overlay
enum SymbolData {
    // MARK: - Tab 1: Full-width (14 rows x 6 cols)

    static let fullWidthRows: [[String]] = [
        ["，", "。", "！", "？", "；", "："],
        ["、", "．", "‧", "…", "～", "·"],
        ["—", "–", "＿", "－", "﹏", "＝"],
        ["「", "」", "『", "』", "（", "）"],
        ["《", "》", "〈", "〉", "【", "】"],
        ["﹁", "﹂", "﹃", "﹄", "〔", "〕"],
        ["［", "］", "｛", "｝", "＂", "＇"],
        ["＠", "＃", "＄", "％", "＆", "＊"],
        ["＋", "＜", "＞", "/", "＼", "｜"],
        ["￠", "￡", "￥", "￦", "＾", "｀"],
        ["❬", "❭", "❰", "❱", "⟨", "⟩"],
        ["⟪", "⟫", "⌈", "⌉", "⌊", "⌋"],
        ["←", "↑", "→", "↓", "⇐", "⇒"],
        ["⇑", "⇓", "⇔", "⇕", "⏎", "↵"],
    ]

    // MARK: - Tab 2: Half-width (8 rows x 6 cols)

    static let halfWidthRows: [[String]] = [
        [",", ".", "!", "?", ";", ":"],
        ["\"", "'", "-", "~", "`", "_"],
        ["(", ")", "[", "]", "{", "}"],
        ["<", ">", "/", "\\", "^", "|"],
        ["@", "#", "$", "%", "&", "*"],
        ["+", "=", "¬", "¦", "¯", "·"],
        ["₹", "€", "£", "¥", "₩", "¢"],
        ["\u{201C}", "\u{201D}", "\u{2018}", "\u{2019}", "«", "»"],
    ]

    // MARK: - Tab 3: Hiragana (8 rows x 6 cols)

    static let hiraganaRows: [[String]] = [
        ["あ", "い", "う", "え", "お", "か"],
        ["き", "く", "け", "こ", "さ", "し"],
        ["す", "せ", "そ", "た", "ち", "つ"],
        ["て", "と", "な", "に", "ぬ", "ね"],
        ["の", "は", "ひ", "ふ", "へ", "ほ"],
        ["ま", "み", "む", "め", "も", "や"],
        ["ゆ", "よ", "ら", "り", "る", "れ"],
        ["ろ", "わ", "を", "ん", "っ", "ー"],
    ]

    // MARK: - Tab 4: Katakana (8 rows x 6 cols)

    static let katakanaRows: [[String]] = [
        ["ア", "イ", "ウ", "エ", "オ", "カ"],
        ["キ", "ク", "ケ", "コ", "サ", "シ"],
        ["ス", "セ", "ソ", "タ", "チ", "ツ"],
        ["テ", "ト", "ナ", "ニ", "ヌ", "ネ"],
        ["ノ", "ハ", "ヒ", "フ", "ヘ", "ホ"],
        ["マ", "ミ", "ム", "メ", "モ", "ヤ"],
        ["ユ", "ヨ", "ラ", "リ", "ル", "レ"],
        ["ロ", "ワ", "ヲ", "ン", "ッ", "ー"],
    ]

    // MARK: - Tab 5: Kaomoji (20 rows x 2 cols)

    static let kaomojiRows: [[String]] = [
        ["( ˶'ᵕ'˶)", "(´・ω・`)"],
        ["(◕‿◕)", "(≧▽≦)"],
        ["(ﾉ◕ヮ◕)ﾉ*:・ﾟ✧", "(╯°□°)╯︵ ┻━┻"],
        ["┬─┬ノ( º _ ºノ)", "¯\\_(ツ)_/¯"],
        ["(ง •_•)ง", "(っ˘ω˘ς)"],
        ["(つ≧▽≦)つ", "(>_<)"],
        ["(T_T)", "(;_;)"],
        ["( ˘ω˘ )", "(⌐■_■)"],
        ["(╥﹏╥)", "(ﾉ´ヮ`)ﾉ*: ・ﾟ"],
        ["(◠‿◠)", "ʕ•ᴥ•ʔ"],
        ["(=^・^=)", "(ΦωΦ)"],
        ["(◕ᴗ◕✿)", "(✿◠‿◠)"],
        ["(≖_≖)", "(⊙_⊙)"],
        ["(☞ﾟヮﾟ)☞", "☜(ﾟヮﾟ☜)"],
        ["(ノಠ益ಠ)ノ", "(´;ω;`)"],
        ["(⌒▽⌒)", "(*^▽^*)"],
        ["(•̀ᴗ•́)و", "(ᵔᴥᵔ)"],
        ["(¬‿¬)", "(✧ω✧)"],
        ["(◔‿◔)", "(ᗒᗣᗕ)"],
        ["(。♥‿♥。)", "(*≧ω≦)"],
    ]

    /// Returns rows for the given category
    static func rows(for category: SymbolCategory) -> [[String]] {
        switch category {
        case .fullWidth: fullWidthRows
        case .halfWidth: halfWidthRows
        case .hiragana: hiraganaRows
        case .katakana: katakanaRows
        case .kaomoji: kaomojiRows
        }
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/text/smartbar/SymbolData.kt
package com.siansiansu.taigikeyboard.ime.text.smartbar

/**
 * Symbol categories for the symbol selection overlay.
 */
enum class SymbolCategory(
    val label: String,
    val columnCount: Int = 6,
    val fontSize: Float = 20f,
) {
    FULL_WIDTH("全形"),
    HALF_WIDTH("半形"),
    HIRAGANA("平仮名"),
    KATAKANA("片仮名"),
    KAOMOJI("顏文字", columnCount = 2, fontSize = 14f),
}

/**
 * Symbol data for the symbol selection overlay.
 *
 * Provides symbol rows for each category in the symbol panel.
 */
object SymbolData {
    /** Full-width symbols (14 rows x 6 cols) */
    val fullWidthRows: List<List<String>> = listOf(
        listOf("，", "。", "！", "？", "；", "："),
        listOf("、", "．", "‧", "…", "～", "·"),
        listOf("—", "–", "＿", "－", "﹏", "＝"),
        listOf("「", "」", "『", "』", "（", "）"),
        listOf("《", "》", "〈", "〉", "【", "】"),
        listOf("﹁", "﹂", "﹃", "﹄", "〔", "〕"),
        listOf("［", "］", "｛", "｝", "＂", "＇"),
        listOf("＠", "＃", "＄", "％", "＆", "＊"),
        listOf("＋", "＜", "＞", "／", "＼", "｜"),
        listOf("￠", "￡", "￥", "￦", "＾", "｀"),
        listOf("❬", "❭", "❰", "❱", "⟨", "⟩"),
        listOf("⟪", "⟫", "⌈", "⌉", "⌊", "⌋"),
        listOf("←", "↑", "→", "↓", "⇐", "⇒"),
        listOf("⇑", "⇓", "⇔", "⇕", "⏎", "↵"),
    )

    /** Half-width symbols (8 rows x 6 cols) */
    val halfWidthRows: List<List<String>> = listOf(
        listOf(",", ".", "!", "?", ";", ":"),
        listOf("\"", "'", "-", "~", "`", "_"),
        listOf("(", ")", "[", "]", "{", "}"),
        listOf("<", ">", "/", "\\", "^", "|"),
        listOf("@", "#", "$", "%", "&", "*"),
        listOf("+", "=", "¬", "¦", "¯", "·"),
        listOf("₹", "€", "£", "¥", "₩", "¢"),
        listOf("\u201C", "\u201D", "\u2018", "\u2019", "«", "»"),
    )

    /** Hiragana (8 rows x 6 cols) */
    val hiraganaRows: List<List<String>> = listOf(
        listOf("あ", "い", "う", "え", "お", "か"),
        listOf("き", "く", "け", "こ", "さ", "し"),
        listOf("す", "せ", "そ", "た", "ち", "つ"),
        listOf("て", "と", "な", "に", "ぬ", "ね"),
        listOf("の", "は", "ひ", "ふ", "へ", "ほ"),
        listOf("ま", "み", "む", "め", "も", "や"),
        listOf("ゆ", "よ", "ら", "り", "る", "れ"),
        listOf("ろ", "わ", "を", "ん", "っ", "ー"),
    )

    /** Katakana (8 rows x 6 cols) */
    val katakanaRows: List<List<String>> = listOf(
        listOf("ア", "イ", "ウ", "エ", "オ", "カ"),
        listOf("キ", "ク", "ケ", "コ", "サ", "シ"),
        listOf("ス", "セ", "ソ", "タ", "チ", "ツ"),
        listOf("テ", "ト", "ナ", "ニ", "ヌ", "ネ"),
        listOf("ノ", "ハ", "ヒ", "フ", "ヘ", "ホ"),
        listOf("マ", "ミ", "ム", "メ", "モ", "ヤ"),
        listOf("ユ", "ヨ", "ラ", "リ", "ル", "レ"),
        listOf("ロ", "ワ", "ヲ", "ン", "ッ", "ー"),
    )

    /** Kaomoji (20 rows x 2 cols) */
    val kaomojiRows: List<List<String>> = listOf(
        listOf("( ˶'ᵕ'˶)", "(´・ω・`)"),
        listOf("(◕‿◕)", "(≧▽≦)"),
        listOf("(ﾉ◕ヮ◕)ﾉ*:・ﾟ✧", "(╯°□°)╯︵ ┻━┻"),
        listOf("┬─┬ノ( º _ ºノ)", "¯\\_(ツ)_/¯"),
        listOf("(ง •_•)ง", "(っ˘ω˘ς)"),
        listOf("(つ≧▽≦)つ", "(>_<)"),
        listOf("(T_T)", "(;_;)"),
        listOf("( ˘ω˘ )", "(⌐■_■)"),
        listOf("(╥﹏╥)", "(ﾉ´ヮ`)ﾉ*: ・ﾟ"),
        listOf("(◠‿◠)", "ʕ•ᴥ•ʔ"),
        listOf("(=^・^=)", "(ΦωΦ)"),
        listOf("(◕ᴗ◕✿)", "(✿◠‿◠)"),
        listOf("(≖_≖)", "(⊙_⊙)"),
        listOf("(☞ﾟヮﾟ)☞", "☜(ﾟヮﾟ☜)"),
        listOf("(ノಠ益ಠ)ノ", "(´;ω;`)"),
        listOf("(⌒▽⌒)", "(*^▽^*)"),
        listOf("(•̀ᴗ•́)و", "(ᵔᴥᵔ)"),
        listOf("(¬‿¬)", "(✧ω✧)"),
        listOf("(◔‿◔)", "(ᗒᗣᗕ)"),
        listOf("(。♥‿♥。)", "(*≧ω≦)"),
    )

    /** Returns rows for the given category */
    fun rows(category: SymbolCategory): List<List<String>> = when (category) {
        SymbolCategory.FULL_WIDTH -> fullWidthRows
        SymbolCategory.HALF_WIDTH -> halfWidthRows
        SymbolCategory.HIRAGANA -> hiraganaRows
        SymbolCategory.KATAKANA -> katakanaRows
        SymbolCategory.KAOMOJI -> kaomojiRows
    }
}
